<a href="https://colab.research.google.com/github/analauraboliveira/Estudos-Aprendizado-de-Maquina/blob/main/Pr%C3%A1tica_1_Ana_Laura.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Centro de Informática
Disciplina: Aprendizado de Máquina e Ciência de Dados
Data: 14/08/2026

Prática 1 – Do dado ao primeiro modelo

Integrantes:
Ana Laura Barboza Oliveira dos Santos

# BLOCO 1 - Conhecendo os dados

In [ ]:
import pandas as pd

url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"
df = pd.read_csv(url)

In [ ]:
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


> **Tarefa 1**

**Qual coluna representa o resultado que queremos prever?**

A coluna que representa o resultado que queremos prever é Survived.

**Quais colunas parecem representar características dos passageiros?**

Algumas características dos passageiros são Pclass, Sex, Age, SibSp, Parch e Fare. Embarked também pode ser considerada, pois registra o embarque do passageiro.

**quais colunas contêm texto?**

As colunas que contêm texto incluem Name, Sex e Embarked.

**Quais colunas contêm números?**

As colunas que contêm números incluem PassengerId, Survived, Pclass, Age, SibSp, Parch e Fare.

In [ ]:
df.shape

(891, 12)

> **Tarefa 2**

**Quantos passageiros estão no conjunto de dados?**

891.

**Quantas colunas existem?**
12.

In [ ]:
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [ ]:
df["Age"]

,Age
0,22.0
1,38.0
2,26.0
3,35.0
4,35.0
...,...
886,27.0
887,19.0
888,NaN
889,26.0


In [ ]:
df[["Name", "Age", "Fare"]]

,Name,Age,Fare
0,"Braund, Mr. Owen Harris",22.0,7.2500
1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",38.0,71.2833
2,"Heikkinen, Miss. Laina",26.0,7.9250
3,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",35.0,53.1000
4,"Allen, Mr. William Henry",35.0,8.0500
...,...,...,...
886,"Montvila, Rev. Juozas",27.0,13.0000
887,"Graham, Miss. Margaret Edith",19.0,30.0000
888,"Johnston, Miss. Catherine Helen ""Carrie""",NaN,23.4500
889,"Behr, Mr. Karl Howell",26.0,30.0000


> **Tarefa 3**

In [ ]:
df[["Name", "Sex", "Age", "Fare"]]

,Name,Sex,Age,Fare
0,"Braund, Mr. Owen Harris",male,22.0,7.2500
1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,71.2833
2,"Heikkinen, Miss. Laina",female,26.0,7.9250
3,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,53.1000
4,"Allen, Mr. William Henry",male,35.0,8.0500
...,...,...,...,...
886,"Montvila, Rev. Juozas",male,27.0,13.0000
887,"Graham, Miss. Margaret Edith",female,19.0,30.0000
888,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,23.4500
889,"Behr, Mr. Karl Howell",male,26.0,30.0000


In [ ]:
df[["Age", "Fare"]]

,Age,Fare
0,22.0,7.2500
1,38.0,71.2833
2,26.0,7.9250
3,35.0,53.1000
4,35.0,8.0500
...,...,...
886,27.0,13.0000
887,19.0,30.0000
888,NaN,23.4500
889,26.0,30.0000


Foram selecionadas as colunas Name, Sex, Age e Fare. Depois, foram selecionadas apenas as colunas Age e Fare.

In [ ]:
df.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


In [ ]:
df["Age"].mean()

np.float64(29.69911764705882)

In [ ]:
df["Age"].median()

28.0

In [ ]:
df["Fare"].mean()

np.float64(32.204207968574636)

In [ ]:
df.isna().sum()

,0
PassengerId,0
Survived,0
Pclass,0
Name,0
Sex,0
Age,177
SibSp,0
Parch,0
Ticket,0
Fare,0


> **Tarefa 4**


As colunas que possuem valores ausentes são Age, Cabin e Embarked.

In [ ]:
df["Sex"].value_counts()

,count
Sex,
male,577
female,314


In [ ]:
df["Pclass"].value_counts()

,count
Pclass,
3,491
1,216
2,184


In [ ]:
df["Survived"].value_counts()

,count
Survived,
0,549
1,342


In [ ]:
df["Pclass"].value_counts().sort_index()

,count
Pclass,
1,216
2,184
3,491


> **Tarefa 5**

**Quantos passageiros sobreviveram?**

342.

**Quantos não sobreviveram?**

549.

**Quantos passageiros pertenciam a cada classe?**

1ª classe: 216 passageiros,
2ª classe: 184 passageiros,
3ª classe: 491 passageiros.

# BLOCO 2 — Manipulando os dados

> **Tarefa 6**

In [ ]:
df[df["Age"] > 50]

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
6,7,0,1,"McCarthy, Mr. Timothy J",male,54.0,0,0,17463,51.8625,E46,S
11,12,1,1,"Bonnell, Miss. Elizabeth",female,58.0,0,0,113783,26.5500,C103,S
15,16,1,2,"Hewlett, Mrs. (Mary D Kingcome)",female,55.0,0,0,248706,16.0000,NaN,S
33,34,0,2,"Wheadon, Mr. Edward H",male,66.0,0,0,C.A. 24579,10.5000,NaN,S
54,55,0,1,"Ostby, Mr. Engelhart Cornelius",male,65.0,0,1,113509,61.9792,B30,C
...,...,...,...,...,...,...,...,...,...,...,...,...
820,821,1,1,"Hays, Mrs. Charles Melville (Clara Jennings Gr...",female,52.0,1,1,12749,93.5000,B69,S
829,830,1,1,"Stone, Mrs. George Nelson (Martha Evelyn)",female,62.0,0,0,113572,80.0000,B28,NaN
851,852,0,3,"Svensson, Mr. Johan",male,74.0,0,0,347060,7.7750,NaN,S
857,858,1,1,"Daly, Mr. Peter Denis",male,51.0,0,0,113055,26.5500,E17,S


In [ ]:
df[df["Pclass"] == 3]

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
5,6,0,3,"Moran, Mr. James",male,NaN,0,0,330877,8.4583,NaN,Q
7,8,0,3,"Palsson, Master. Gosta Leonard",male,2.0,3,1,349909,21.0750,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...,...
882,883,0,3,"Dahlberg, Miss. Gerda Ulrika",female,22.0,0,0,7552,10.5167,NaN,S
884,885,0,3,"Sutehall, Mr. Henry Jr",male,25.0,0,0,SOTON/OQ 392076,7.0500,NaN,S
885,886,0,3,"Rice, Mrs. William (Margaret Norton)",female,39.0,0,5,382652,29.1250,NaN,Q
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.4500,NaN,S


In [ ]:
df[(df["Sex"] == "female") & (df["Pclass"] == 1)]

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
11,12,1,1,"Bonnell, Miss. Elizabeth",female,58.0,0,0,113783,26.5500,C103,S
31,32,1,1,"Spencer, Mrs. William Augustus (Marie Eugenie)",female,NaN,1,0,PC 17569,146.5208,B78,C
52,53,1,1,"Harper, Mrs. Henry Sleeper (Myna Haxtun)",female,49.0,1,0,PC 17572,76.7292,D33,C
...,...,...,...,...,...,...,...,...,...,...,...,...
856,857,1,1,"Wick, Mrs. George Dennick (Mary Hitchcock)",female,45.0,1,1,36928,164.8667,NaN,S
862,863,1,1,"Swift, Mrs. Frederick Joel (Margaret Welles Ba...",female,48.0,0,0,17466,25.9292,D17,S
871,872,1,1,"Beckwith, Mrs. Richard Leonard (Sallie Monypeny)",female,47.0,1,1,11751,52.5542,D35,S
879,880,1,1,"Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)",female,56.0,0,1,11767,83.1583,C50,C


In [ ]:
df[df["Fare"] > 100]

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
27,28,0,1,"Fortune, Mr. Charles Alexander",male,19.00,3,2,19950,263.0000,C23 C25 C27,S
31,32,1,1,"Spencer, Mrs. William Augustus (Marie Eugenie)",female,NaN,1,0,PC 17569,146.5208,B78,C
88,89,1,1,"Fortune, Miss. Mabel Helen",female,23.00,3,2,19950,263.0000,C23 C25 C27,S
118,119,0,1,"Baxter, Mr. Quigg Edmond",male,24.00,0,1,PC 17558,247.5208,B58 B60,C
195,196,1,1,"Lurette, Miss. Elise",female,58.00,0,0,PC 17569,146.5208,B80,C
215,216,1,1,"Newell, Miss. Madeleine",female,31.00,1,0,35273,113.2750,D36,C
258,259,1,1,"Ward, Miss. Anna",female,35.00,0,0,PC 17755,512.3292,NaN,C
268,269,1,1,"Graham, Mrs. William Thompson (Edith Junkins)",female,58.00,0,1,PC 17582,153.4625,C125,S
269,270,1,1,"Bissette, Miss. Amelia",female,35.00,0,0,PC 17760,135.6333,C99,S
297,298,0,1,"Allison, Miss. Helen Loraine",female,2.00,1,2,113781,151.5500,C22 C26,S


>**Tarefa 7**

In [ ]:
df.sort_values("Fare", ascending=False)[["Name", "Pclass", "Fare"]].head()

,Name,Pclass,Fare
679,"Cardeza, Mr. Thomas Drake Martinez",1,512.3292
258,"Ward, Miss. Anna",1,512.3292
737,"Lesurer, Mr. Gustave J",1,512.3292
88,"Fortune, Miss. Mabel Helen",1,263.0000
438,"Fortune, Mr. Mark",1,263.0000


**Resposta:** Mr. Thomas Drake Martinez Cardeza, Miss. Anna Ward, Mr. Gustave J Lesurer, Miss. Mabel Helen Fortune e Mr. Mark Fortune.

In [ ]:
df["Familia"] = df["SibSp"] + df["Parch"]

In [ ]:
df[["SibSp", "Parch", "Familia"]].head()

,SibSp,Parch,Familia
0,1,0,1
1,1,0,1
2,0,0,0
3,1,0,1
4,0,0,0


> **Tarefa 8**

In [ ]:
df["Sozinho"] = (df["SibSp"] + df["Parch"] == 0)

In [ ]:
df["Sozinho"].value_counts()

,count
Sozinho,
True,537
False,354


**Resposta**: 537 passageiros viajavam sozinhos e 354 viajavam acompanhados de pelo menos um familiar.

> **Tarefa 9**

In [ ]:
df.groupby("Sex")["Survived"].mean()

,Survived
Sex,
female,0.742038
male,0.188908


In [ ]:
df.groupby("Pclass")["Survived"].mean()

,Survived
Pclass,
1,0.629630
2,0.472826
3,0.242363


In [ ]:
df.groupby(["Pclass", "Sex"])["Survived"].mean()

Pclass  Sex   
1       female    0.968085
        male      0.368852
2       female    0.921053
        male      0.157407
3       female    0.500000
        male      0.135447
Name: Survived, dtype: float64

# BLOCO 3 — PRIMEIRO MODELO DE MACHINE LEARNING

In [ ]:
dados = df[[
    "Survived",
    "Pclass",
    "Sex",
    "Age",
    "SibSp",
    "Parch",
    "Fare"
]].copy()

In [ ]:
dados["Sex"] = (dados["Sex"] == "female").astype(int)

In [ ]:
dados.head()

,Survived,Pclass,Sex,Age,SibSp,Parch,Fare
0,0,3,0,22.0,1,0,7.2500
1,1,1,1,38.0,1,0,71.2833
2,1,3,1,26.0,0,0,7.9250
3,1,1,1,35.0,1,0,53.1000
4,0,3,0,35.0,0,0,8.0500


In [ ]:
dados["Age"] = dados["Age"].fillna(dados["Age"].median())

In [ ]:
dados.isna().sum()

,0
Survived,0
Pclass,0
Sex,0
Age,0
SibSp,0
Parch,0
Fare,0


In [ ]:
X = dados.drop(columns="Survived")
y = dados["Survived"]

In [ ]:
X.head()

,Pclass,Sex,Age,SibSp,Parch,Fare
0,3,0,22.0,1,0,7.2500
1,1,1,38.0,1,0,71.2833
2,3,1,26.0,0,0,7.9250
3,1,1,35.0,1,0,53.1000
4,3,0,35.0,0,0,8.0500


In [ ]:
y.head()

,Survived
0,0
1,1
2,1
3,1
4,0


In [ ]:
from sklearn.model_selection import train_test_split

X_treino, X_teste, y_treino, y_teste = train_test_split(
    X,
    y,
    test_size=0.30,
    random_state=42,
    stratify=y
)

In [ ]:
X_treino.shape

(623, 6)

In [ ]:
X_teste.shape

(268, 6)

In [ ]:
from sklearn.neighbors import KNeighborsClassifier

In [ ]:
knn = KNeighborsClassifier(n_neighbors=5)

In [ ]:
knn.fit(X_treino, y_treino)

KNeighborsClassifier()

In [ ]:
previsoes = knn.predict(X_teste)

In [ ]:
previsoes[:10]

array([0, 0, 0, 0, 0, 1, 0, 0, 0, 1])

In [ ]:
y_teste.values[:10]

array([0, 0, 0, 1, 1, 0, 0, 1, 0, 1])

> **Tarefa 10**

**Quantos dos dez primeiros exemplos foram classificados corretamente?**

Dos 10 primeiros exemplos do conjunto de teste, 6 foram classificados corretamente.

> **Tarefa 11**

In [ ]:
knn.score(X_teste, y_teste)

0.6791044776119403

In [ ]:
(previsoes == y_teste.values).mean()

np.float64(0.6791044776119403)

In [ ]:
knn.score(X_treino, y_treino)

0.8089887640449438

**Resposta:** A acurácia obtida no conjunto de teste foi aproximadamente 67,91%. No conjunto de treinamento, a acurácia foi aproximadamente 80,90%.

In [ ]:
for k in [1, 3, 5, 7, 10]:
    modelo = KNeighborsClassifier(n_neighbors=k)
    modelo.fit(X_treino, y_treino)
    acuracia = modelo.score(X_teste, y_teste)
    print(k, acuracia)

1 0.7014925373134329
3 0.6753731343283582
5 0.6791044776119403
7 0.6753731343283582
10 0.6865671641791045


> **Tarefa 12**

**Qual k apresentou a maior acurácia?**

k = 1, com aproximadamente 70,15%.

**Qual apresentou a menor?**

k = 3 e k = 7, ambos com aproximadamente 67,54%.

# BLOCO 4 - Fechando a prática

1. **Quantas instâncias e atributos possui o conjunto de dados?**

O conjunto de dados possui 891 instâncias (passageiros) e 12 atributos (colunas).


2. **Quais atributos foram utilizados como variáveis de entrada (X) e qual atributo foi utilizado como variável-alvo (y)?**

As variáveis de entrada X foram Pclass, Sex, Age, SibSp, Parch e Fare. A variável-alvo y foi Survived.

3. **Qual é a função da divisão dos dados em conjuntos de treinamento e teste?**

A divisão serve para utilizar uma parte dos dados para treinar o modelo e outra parte para testar e avaliar seu desempenho em dados que não foram utilizados/vistos durante o treinamento.

4. **O que acontece quando o modelo KNN recebe uma nova instância para realizar uma classificação?**

O KNN procura os k vizinhos mais próximos da nova instância e utiliza as classes desses vizinhos para determinar a classificação da nova instância.

5. **Qual foi a acurácia obtida pelo modelo no conjunto de teste?**

A acurácia obtida pelo modelo no conjunto de teste foi de aproximadamente 67,91% (0.6791).

6. **Por que o conjunto de teste não deve ser utilizado durante o treinamento do modelo?**

Porque o conjunto de teste deve conter dados que o modelo não viu durante o treinamento, permitindo avaliar corretamente seu desempenho em novos exemplos.

7. **Qual foi a principal dúvida ou dificuldade encontrada durante a atividade? **


Minha principal dificuldade foi entender alguns dos comandos do pandas e como utilizá-los para filtrar, organizar e analisar os dados da tabela.

> Fim da atividade